lib/driver.js

```diff
@@ -184,8 +194,29 @@ class BaseDriver extends MobileJsonWireProtocol {
 
     if (!_.contains(validStrats, strategy)) {
       throw new errors.InvalidSelectorError(`Locator Strategy "${strategy}" is not supported for this session`);
+    }
+  }
+
+  /*
+   * Restart the session with the original caps,
+   * preserving the timeout config.
+   */
+  async reset () {
+    log.debug("Resetting app mid-session");
+    log.debug("Running generic full reset");
+    let currentConfig = {};
+    for (let k of ['implicitWaitMs', 'newCommandTimeoutMs', 'sessionId']) {
+      currentConfig[k] = this[k];
+    }
+    await this.deleteSession();
+    log.debug("Restarting app");
+    await this.createSession(this.caps);
+
+    for (let [k,v] of _.pairs(currentConfig)) {
+      this[k] = v;
```

Comment:
I wonder if there's a bug here when `createSession` begins proxying to something else (like Chromedriver or Selendroid), because the JWProxy object will be told that our current sessionId is whatever new one `createSession` comes up with, rather than the one we are trying to retain in `currentConfig`. We should test this case and make sure it's not a bug. If it is, we might want to be able to pass in a sessionId to `createSession` so it uses that one instead of coming up with a new one.